### PR TITLE
More clarifications on mutating Widget's children

### PR DESCRIPTION
**Pull request by liyuqian**
This summarizes what Ian and Yuqian discussed in the chat room.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
   /// If this list is going to be mutated, it is usually wise to put [Key]s on
   /// the widgets, so that the framework can match old configurations to new
   /// configurations and maintain the underlying render objects.
+  ///
+  /// Here's a common example where one mutates the children but forgets to set
+  /// a different key. In such case, the mutated children may not be reflected.
+  /// ```
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: \`\`\`dart

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  /// class SomeWidgetState extends ... {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: fill in the `...`

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  ///
+  ///     void someHandler() {
+  ///         setState(() {
+  ///             _children.add(...);
+  ///         });
+  ///     }
+  ///
+  ///     Widget build(...) {
+  ///         // A changing key is needed if someHandler modifies _children.
+  ///         return Row(children: _children/*, key: missing_key*/);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: woah no, that won't work. you'll be blowing away the entire Row each time if you do this, that's very expensive.

You want to set the keys on each of the _children, and you want to very explicitly _not_ have a `List` that you create outside of the `build`.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  ///     }
+  /// }
+  /// ```
+  ///
+  /// Also, a [Widget] in Flutter is immutable, so directly modifying the
+  /// [children] such as `someMultiChildRenderObjectWidget.children.add(...)`
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: or as in the exact way that the code above does

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  /// is a violation to Flutter's contract. Although no warning or error message
+  /// is printed out for such case right now, we advice against such usages to
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: advice -> advise

File: packages/flutter/lib/src/widgets/framework.dart [hunk 6]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  ///     Widget build(...) {
+  ///         // A changing key is needed if someHandler modifies _children.
+  ///         return Row(children: _children/*, key: missing_key*/);
+  ///     }
+  /// }
+  /// ```
+  ///
+  /// Also, a [Widget] in Flutter is immutable, so directly modifying the
+  /// [children] such as `someMultiChildRenderObjectWidget.children.add(...)`
+  /// is a violation to Flutter's contract. Although no warning or error message
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: "violation of"

File: packages/flutter/lib/src/widgets/framework.dart [hunk 7]
```diff
@@ -1732,6 +1732,35 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  /// is printed out for such case right now, we advice against such usages to
+  /// avoid undefined behaviors.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: "advise against" is probably too weak, and "violation of Flutter's contract" is probably too strong-sounding. I would just say that passing the same actual List object each time will result in incorrect behaviour, and that a new List should be provided each time the class is built.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 8]
```diff
@@ -1729,9 +1729,36 @@ abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
...
+  /// configurations to new configurations and maintain the underlying render
+  /// objects.
+  ///
+  /// Also, a [Widget] in Flutter is immutable, so directly modifying the
+  /// [children] such as `someMultiChildRenderObjectWidget.children.add(...)` or
+  /// as the example code below will result in incorrect behaviors. Whenever the
+  /// children list is modified, a new list object should be provided.
+  ///
+  /// ```dart
+  /// class SomeWidgetState extends State<SomeWidget> {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart]: Instead if showing a "bad" example can we show a good example that shows people how to do it right? I believe that will be more useful.

Comment (liyuqian) [reply packages/flutter/lib/src/widgets/framework.dart]: Sure, added a positive example. How does it look?

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**